ethereum/weaksubjectivity/src/main/java/tech/pegasys/teku/weaksubjectivity/WeakSubjectivityValidator.java

```diff
@@ -68,6 +110,35 @@ public void validateLatestFinalizedCheckpoint(
     }
   }
 
+  public boolean isBlockValid(
+      final SignedBeaconBlock block, ForkChoiceStrategy forkChoiceStrategy) {
+    if (maybeWsCheckpoint.isEmpty()) {
+      return true;
+    }
+    final Checkpoint wsCheckpoint = maybeWsCheckpoint.get();
+
+    UInt64 blockEpoch = compute_epoch_at_slot(block.getSlot());
+    boolean blockAtEpochBoundary = compute_start_slot_at_epoch(blockEpoch).equals(block.getSlot());
+    if (isWSCheckpointEpoch(blockEpoch) && blockAtEpochBoundary) {
+      // Block is at ws checkpoint slot - so it must match the ws checkpoint block
+      return isWSCheckpointBlock(block);
+    } else if (isWSCheckpointBlock(block)) {
+      // The block is the checkpoint
+      return true;
+    } else if (blockEpoch.isGreaterThanOrEqualTo(wsCheckpoint.getEpoch())) {
+      // If the block is at or past the checkpoint, the wsCheckpoint must be an ancestor
+      final Optional<Bytes32> ancestor =
+          get_ancestor(
+              forkChoiceStrategy, block.getParent_root(), wsCheckpoint.getEpochStartSlot());
+      // If ancestor is not present, the chain must have moved passed the wsCheckpoint
+      return ancestor.map(a -> a.equals(wsCheckpoint.getRoot())).orElse(true);
```

Comment:
nit: It might be nice to log a warning if the chain moved passed the wsCheckpoint since we can't verify it anymore.

Reply:
Yeah - the thought crossed my mind as well.  But we should be gating all imports with these checks, and validating consistency with the wsCheckpoint at startup, which should prevent us from following an inconsistent chain.  The only way we hit this is if we don't have the parent of the block being imported (which will get validated later), or the wsCheckpoint has fallen off behind the latest finalized block. Given that this is an expected case and could get really noisey, I thought it was best to just move on here.